adapters/startio/startio.go

```diff
@@ -0,0 +1,179 @@
+		return nil, []error{err}
+	}
+
+	validImpressions, err := getValidImpressions(requestCopy.Imp)
+	if err != nil {
+		return nil, []error{err}
+	}
+	for i := range validImpressions {
+		requestCopy.Imp = []openrtb2.Imp{validImpressions[i]}
+
+		requestBody, err := jsonutil.Marshal(requestCopy)
+		if err != nil {
+			errors = append(errors, fmt.Errorf("imp[%d]: failed to marshal request: %w", i, err))
+			continue
+		}
+
+		requestData := &adapters.RequestData{
+			Method:  http.MethodPost,
+			Uri:     adapter.endpoint,
+			Body:    requestBody,
+			Headers: buildRequestHeaders(),
+			ImpIDs:  []string{validImpressions[i].ID},
+		}
+
+		requests = append(requests, requestData)
+	}
+
+	return requests, errors
+}
+
+func (adapter *StartioAdapter) MakeBids(internalRequest *openrtb2.BidRequest, externalRequest *adapters.RequestData, response *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	if response.StatusCode == http.StatusNoContent {
+		return nil, nil
+	} else if response.StatusCode != http.StatusOK {
+		return nil, []error{wrapReqError(fmt.Sprintf("Unexpected status code: %d.", response.StatusCode))}
+	}
+
+	if err := adapters.CheckResponseStatusCodeForErrors(response); err != nil {
+		return nil, []error{err}
+	}
+
+	var bidResponse openrtb2.BidResponse
+	if err := jsonutil.Unmarshal(response.Body, &bidResponse); err != nil {
+		return nil, []error{wrapReqError(fmt.Sprintf("failed to unmarshal response body: %v", err))}
+	}
+
+	var errs []error
+	bidderResponse := adapters.NewBidderResponseWithBidsCapacity(len(bidResponse.SeatBid))
+
+	for i := range bidResponse.SeatBid {
+		for j := range bidResponse.SeatBid[i].Bid {
+			bid := &bidResponse.SeatBid[i].Bid[j]
+			bidType, err := getMediaTypeForBid(*bid)
+
+			if err != nil {
+				errs = append(errs, err)
+			} else {
+				bidderResponse.Bids = append(bidderResponse.Bids, &adapters.TypedBid{
+					Bid:     bid,
+					BidType: bidType,
+				})
+			}
+		}
+	}
+
+	return bidderResponse, errs
+}
+
+func validateRequest(request openrtb2.BidRequest) error {
+	if !isSupportedCurrency(request.Cur) {
+		return wrapReqError("unsupported currency: only USD is accepted")
+	}
+
+	if !hasSiteOrAppID(request) {
+		return wrapReqError("request must contain either site.id or app.id")
+	}
+
+	return nil
+}
+
+func getValidImpressions(imps []openrtb2.Imp) ([]openrtb2.Imp, error) {
+	var validImpressions []openrtb2.Imp
+
+	for _, imp := range imps {
+		imp.Audio = nil
+		hasValidMedia := imp.Banner != nil || imp.Video != nil || imp.Native != nil
+		if hasValidMedia {
+			validImpressions = append(validImpressions, imp)
+		}
+	}
+
+	if len(validImpressions) == 0 {
+		return nil, wrapReqError("invalid bid request: at least one banner, video, or native impression is required.")
+	}
+
+	return validImpressions, nil
+}
+
+func hasSiteOrAppID(req openrtb2.BidRequest) bool {
+	return (req.Site != nil && req.Site.ID != "") || (req.App != nil && req.App.ID != "")
+}
```

Comment:
I just want to point that there is already some site and app validation being performed by PBS core. Please see `endpoints/openrtb2/auction.go#validateSite` and `endpoints/openrtb2/auction.go#validateApp` to determine if the validation being performed there suits your needs. If it does, you can remove this code as your adapter would never be called if that PBS core validation fails.